cpp/src/arrow/array/array_test.cc

```diff
@@ -499,6 +501,24 @@ TEST_F(TestArray, TestMakeArrayOfNull) {
       }
     }
   }
+
+  for (int64_t length : {0, 16}) {
+    ARROW_SCOPED_TRACE("length = ", length, " (required fields)");
+
+    auto req = [](auto type) { return field("", std::move(type), /*nullable=*/false); };
+
+    // union with no nullable fields cannot represent a null
+    ASSERT_RAISES(Invalid, MakeArrayOfNull(dense_union({req(int8())}), length));
+
+    // struct with no nullable fields has a top level bitmap and can mask them
+    ASSERT_OK_AND_ASSIGN(auto s, MakeArrayOfNull(struct_({req(int8())}), length));
+    ASSERT_OK(s->ValidateFull());
+
+    // dictionary with non-nullable indices can use a 1-long dict of null
+    ASSERT_OK_AND_ASSIGN(
+        s, MakeArrayOfNull(struct_({req(dictionary(int8(), int8()))}), length));
+    ASSERT_OK(s->ValidateFull());
```

Comment:
Do we also want to test run-end encoded types with non-nullable value type?

Reply:
will do